src/components/legend/draw.js

```diff
@@ -392,24 +393,57 @@ function drawTexts(g, gd) {
                 this.text(text)
                     .call(textLayout);
 
+                var origText = text;
+
                 if(!this.text()) text = ' \u0020\u0020 ';
 
-                var fullInput = legendItem.trace._fullInput || {},
-                    astr;
+                var i, transforms, direction;
+                var fullInput = legendItem.trace._fullInput || {};
+                var needsRedraw = false;
+                var update = {};
 
                 // N.B. this block isn't super clean,
                 // is unfortunately untested at the moment,
                 // and only works for for 'ohlc' and 'candlestick',
                 // but should be generalized for other one-to-many transforms
                 if(['ohlc', 'candlestick'].indexOf(fullInput.type) !== -1) {
-                    var transforms = legendItem.trace.transforms,
-                        direction = transforms[transforms.length - 1].direction;
+                    transforms = legendItem.trace.transforms;
+                    direction = transforms[transforms.length - 1].direction;
+
+                    update[direction + '.name'] = text;
+                } else if(fullInput.transforms) {
+                    for(i = fullInput.transforms.length - 1; i >= 0; i--) {
+                        if(fullInput.transforms[i].type === 'groupby') {
+                            break;
+                        }
+                    }
+
+                    var carr = Lib.keyedContainer(fullInput, 'transforms[' + i + '].groupnames', 'group', 'name');
+
+                    if(BLANK_STRING_REGEX.test(origText)) {
+                        carr.remove(legendItem.trace._group);
+                        needsRedraw = true;
+                    } else {
+                        carr.set(legendItem.trace._group, text);
+                    }
+
+                    update = carr.constructUpdate();
+                } else {
+                    update.name = text;
+                }
+
+                var p = Plotly.restyle(gd, update, traceIndex);
 
-                    astr = direction + '.name';
+                // If a groupby label is deleted, it seems like we need another redraw in order
+                // to restore the label. Otherwise it simply sets this property and the blank
+                // string is retained.
+                if(needsRedraw) {
+                    p = p.then(function() {
+                        return Plotly.redraw(gd);
```

Comment:
Hmm. We shouldn't have to do this. We should fix [`_restyle`](https://github.com/plotly/plotly.js/blob/b2e54861dfe9960a7bb447f0a622d4198a4e22c7/src/plot_api/plot_api.js#L1295) so that it leads to the correct update.